llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,322 @@
+      // Delete the extra ones in the new last page.
+      uintptr_t PagePtr = PageToDataIdx[NewLastPage];
+      if (PagePtr != InvalidPage) {
+        T *Page = reinterpret_cast<T *>(PagePtr);
+        // If the new size and the old size are on the same page, we need to
+        // delete only the elements between the new size and the old size.
+        // Otherwise we need to delete all the remaining elements in the page.
+        std::size_t OldPage = (Size - 1) / PAGE_SIZE;
+        std::size_t NewPage = (NewSize - 1) / PAGE_SIZE;
+        std::size_t LastPageElements =
+            OldPage == NewPage ? Size % PAGE_SIZE : PAGE_SIZE;
+        for (std::size_t J = NewSize % PAGE_SIZE; J < LastPageElements; ++J)
+          Page[J].~T();
+      }
+      PageToDataIdx.resize(NewLastPage + 1);
+    }
+    Size = NewSize;
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages.
+    if (Size <= capacity())
+      return;
+    // The number of pages to allocate. The Remainder is calculated
+    // for the case in which the NewSize is not a multiple of PAGE_SIZE.
+    // In that case we need one more page.
+    std::size_t Pages = Size / PAGE_SIZE;
+    std::size_t Remainder = Size % PAGE_SIZE;
+    if (Remainder != 0)
+      Pages += 1;
+    assert(Pages > PageToDataIdx.size());
+    // We use InvalidPage to indicate that a page has not been allocated yet.
+    // This cannot be 0, because 0 is a valid page id.
+    // We use InvalidPage instead of a separate bool to avoid wasting space.
+    PageToDataIdx.resize(Pages, InvalidPage);
+  }
+
+  // Return true if the vector is empty
+  [[nodiscard]] bool empty() const { return Size == 0; }
+
+  /// Clear the vector, i.e. clear the allocated pages, the whole page
+  /// lookup index and reset the size.
+  void clear() {
+    Size = 0;
+    // If we own the allocator, simply reset it, otherwise we
+    // deallocate the pages one by one.
+    if (Allocator.getInt() == true)
+      Allocator.getPointer()->Reset();
+    else
+      for (uintptr_t Page : PageToDataIdx)
+        Allocator.getPointer()->Deallocate(reinterpret_cast<T *>(Page));
+
+    PageToDataIdx.clear();
+  }
+
+  // Iterator on all the elements of the vector
+  // which have actually being constructed.
+  class MaterialisedIterator {
+    PagedVector const *PV;
+    std::size_t ElementIdx;
+
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = T;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T *;
+    using reference = T &;
+
+    MaterialisedIterator(PagedVector const *PV, std::size_t ElementIdx)
+        : PV(PV), ElementIdx(ElementIdx) {}
+
+    // When incrementing the iterator, we skip the elements which have not
+    // been materialised yet.
+    MaterialisedIterator &operator++() {
+      while (ElementIdx < PV->Size)
+        if (PV->PageToDataIdx[ElementIdx++ / PAGE_SIZE] != InvalidPage)
+          break;
```

Comment:
May as well skip whole pages rather than elements here.
```suggestion
      ++ElementIdx;
      if (ElementIdx % PAGE_SIZE == 0) {
        while (ElementIdx < PV->Size && PV->PageToDataIdx[ElementIdx / PAGE_SIZE] == InvalidPage)
          ElementIdx += PAGE_SIZE;
        if (ElementIdx > PV->Size)
          ElementIdx = PV->Size;
      }
```